Add trailing slash to qute://help pages

deewakar opened the pull request:
qute://help (without the trailing slash) opens about:blank page due to path being empty. This fixes the issue by redirecting to the root path when path is empty.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3215)
<!-- Reviewable:end -->


edit by @The-Compiler: Fixes #3161 (I think this needs to be in the body for GitHub to auto-close the issue)

The-Compiler commented:
Can you please also add a test to `tests/end2end/features/qutescheme.feature`? See `Scenario: Opening link with qute:help` for a similar test there.

deewakar commented:
> from a quick look, this is because they do Then qute://log?level=info should be loaded, but the path now changed to qute://log/?level=info.

No, currently it loads only qute://log/ i.e. query parameters are missing. I changed the following to remedy that.
``` diff
diff --git a/qutebrowser/browser/qutescheme.py b/qutebrowser/browser/qutescheme.py
index 518012f10..0c4f611f5 100644
--- a/qutebrowser/browser/qutescheme.py
+++ b/qutebrowser/browser/qutescheme.py
@@ -138,6 +138,7 @@ def data_for_url(url):
     """
     path = url.path()
     host = url.host()
+    query = url.query()
     # A url like "qute:foo" is split as "scheme:path", not "scheme:host".
     log.misc.debug("url: {}, path: {}, host {}".format(
         url.toDisplayString(), path, host))
@@ -152,6 +153,8 @@ def data_for_url(url):
             new_url.setHost(path)
 
         new_url.setPath('/')
+        if query:
+            new_url.setQuery(query)
         if new_url.host():  # path was a valid host
             raise Redirect(new_url)
 
```

Also made these changes to tests/end2end/features/qutescheme.feature

``` diff
diff --git a/tests/end2end/features/qutescheme.feature b/tests/end2end/features/qutescheme.feature
index 8b1a156a0..662efdc26 100644
--- a/tests/end2end/features/qutescheme.feature
+++ b/tests/end2end/features/qutescheme.feature
@@ -147,12 +147,12 @@ Feature: Special qute:// pages
 
     Scenario: Running :pyeval
         When I run :debug-pyeval 1+1
-        And I wait until qute://pyeval is loaded
+        And I wait until qute://pyeval/ is loaded
         Then the page should contain the plaintext "2"
 
     Scenario: Causing exception in :pyeval
         When I run :debug-pyeval 1/0
-        And I wait until qute://pyeval is loaded
+        And I wait until qute://pyeval/ is loaded
         Then the page should contain the plaintext "ZeroDivisionError"
 
     Scenario: Running :pyeval with --quiet
@@ -166,7 +166,7 @@ Feature: Special qute:// pages
         And I run :message-warning the-warning-message
         And I run :message-info the-info-message
         And I run :messages
-        Then qute://log?level=info should be loaded
+        Then qute://log/?level=info should be loaded
         And the error "the-error-message" should be shown
         And the warning "the-warning-message" should be shown
         And the page should contain the plaintext "the-error-message"
@@ -178,7 +178,7 @@ Feature: Special qute:// pages
         And I run :message-warning the-warning-message
         And I run :message-info the-info-message
         And I run :messages warning
-        Then qute://log?level=warning should be loaded
+        Then qute://log/?level=warning should be loaded
         And the error "the-error-message" should be shown
         And the warning "the-warning-message" should be shown
         And the page should contain the plaintext "the-error-message"
@@ -190,7 +190,7 @@ Feature: Special qute:// pages
         And I run :message-warning the-warning-message
         And I run :message-info the-info-message
         And I run :messages info
-        Then qute://log?level=info should be loaded
+        Then qute://log/?level=info should be loaded
         And the error "the-error-message" should be shown
         And the warning "the-warning-message" should be shown
         And the page should contain the plaintext "the-error-message"
```
Then I ran ```tox -e py36-pyqt59 tests/end2end/features/test_qutescheme_bdd.py```
Some tests still failed, here's the summary

```
=========================== short test summary info ============================
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_opening_link_with_qutehelp
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_opening_a_link_with_qutehelp
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_running_pyeval
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_causing_exception_in_pyeval
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_running_pyeval_with_quiet
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_messages_without_level
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_showing_messages_of_type_warning_or_greater
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_showing_messages_of_type_info_or_greater
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_using_qutelog_directly
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_using_quteplainlog_directly
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_open_quteversion
============================== 0 tests deselected ==============================
========= 11 failed, 10 passed, 3 skipped, 1 xfailed in 79.11 seconds ==========

```
Here's the complete log of that test run -> https://gist.github.com/deewakar/ab8bedcbf9e2d440b12aa30bd1e82a74

The-Compiler commented:
Hmm, I don't understand why you'd get all those failures, they mostly seem to be because `about:blank` isn't loaded? 😕 

All I get is:

```
FAIL tests/end2end/features/test_qutescheme_bdd.py::test_opening_a_link_with_qutehelp
```

which is fixed by:

```diff
diff --git a/tests/end2end/features/qutescheme.feature b/tests/end2end/features/qutescheme.feature
index 662efdc26..43558cf60 100644
--- a/tests/end2end/features/qutescheme.feature
+++ b/tests/end2end/features/qutescheme.feature
@@ -58,7 +58,7 @@ Feature: Special qute:// pages
     Scenario: Opening a link with qute://help
         When the documentation is up to date
         And I run :tab-only
-        And I open qute://help
+        And I open qute://help without waiting
         And I wait until qute://help/ is loaded
         And I hint with args "links normal" and follow a
         Then qute://help/quickstart.html should be loaded
```

I suggest you push all those fixes to your PR and we see what AppVeyor/Travis think :wink:

The-Compiler commented:
Thank you!
